jobserver fd_check: Allow sockets.

xim opened the pull request:
I agree with the creator of issue #6 that jobserver fds should be allowed to be sockets as well as pipes. A project I'm working on uses a jobserver implemented over Unix domain sockets.

xim committed:
jobserver fd_check: Allow sockets and char devices

Any fd with compatible semantics can be used for jobserver tokens, not
just pipes. Multiple jobserver implementations use other fd types
already. Other client implementations don't require pipes either, like
those in `make` and `ninja`.

Actually fixes issue #6

Co-Authored-By: CertainLach <6235312+CertainLach@users.noreply.github.com>

xim committed:
jobserver fd_check: Allow sockets.

I agree with the creator of issue #6 that jobserver fds should be
allowed to be sockets as well as pipes. A project I'm working on
uses a jobserver implemented over Unix domain sockets.

xim committed:
fixup! jobserver fd_check: Allow sockets.

xim commented:
Can add some details ; I already have a piece of software running that creates a token pool across all builds on the local machine using an abstract unix domain socket. It works well with GNU Make, Ninja and gcc.

xim commented:
I'm more than willing to contribute, and to consider alternative approaches.

NobodyXu commented on src/error.rs:
```diff
@@ -48,7 +48,7 @@ impl FromEnvError {
             FromEnvErrorInner::CannotOpenFd(..) => FromEnvErrorKind::CannotOpenFd,
             FromEnvErrorInner::NegativeFd(..) => FromEnvErrorKind::NegativeFd,
-            FromEnvErrorInner::NotAPipe(..) => FromEnvErrorKind::NotAPipe,
+            FromEnvErrorInner::NotAPipeOrSocket(..) => FromEnvErrorKind::NotAPipeOrSocket,
```
This would probably be an undesired breaking change, don't think we can rename this enum variant which is unfortunate

Could add a `NotASocket`

xim replied:
I checked all uses in cargo/rust, this value was never explicitly referenced.

weihanglo replied:
It is better to avoid MSRV bump. I would suggest adding new variant and leave the old one there. We already have `#[non_exhaustive]` so should be fine adding new variants.

xim replied:
Kept the old public variant and marked it as deprecated

NobodyXu reviewed (commented):
I'm not a maintainer of jobserver though I did have some contributions and subscribe to updates to jobserver closely and is a maintainer of cc, which use jobserver, so take my comment with a grain of salt.

And I'm not sure if I should merge the PR here on behalf of the maintainer, so probably not unless nobody has time to review it

CertainLach commented on src/unix.rs:
```diff
@@ -495,7 +495,14 @@ unsafe fn fd_check(fd: c_int, check_pipe: bool) -> Result<(), FromEnvErrorInner>
+            s_ifsock = libc::S_IFSOCK as _;
+            if stat.st_mode & s_ifsock == s_ifsock {
+                return Ok(());
+            }
```
There are jobserver implementations that use CUSE (character device in userspace, S_IFCHR), e.g

https://gitweb.gentoo.org/proj/steve.git
https://codeberg.org/amonakov/guildmaster
https://github.com/NixOS/nixpkgs/pull/314888

Would be great if this PR permitted them too:

```suggestion
            let fmt = |v| {
                #[allow(unused_assignments)]
                let mut s_ifmt = stat.st_mode;
                s_ifmt = v as _;
                s_ifmt
            };
            
            let fmt = stat.st_mode & fmt(libc::S_IFMT);
            if fmt == fmt(libc::S_IFIFO) || fmt == fmt(libc::S_IFSOCK) || fmt == fmt(libc::S_IFCHR)  {
                return Ok(());
            }
            Err(FromEnvErrorInner::NotAPipe(fd, None))
```

xim replied:
I did that, with some minor changes. Added you as co-author. Also renamed the error enum and text to be more generic.

weihanglo commented on src/unix.rs:
Is it possible we have tests for the new behavior including successful paths and failing paths?

xim replied:
Sure!

xim committed:
jobserver fd_check: Add tests for supported fd types

Pipes, sockets and character devices are supported.

Regular files are not supported.

Additionally, check_pipe=false skips the type check.

weihanglo reviewed (commented):
Thanks for the update. Would you mind rebasing onto master so the CI should pass?
